internal/service/apis.go

```diff
@@ -24,5 +27,64 @@ type SendOptions struct {
 // CryptoAPI provides methods for getting crypto rates that are used in CryptoService and
 // implemented in external packages.
 type CryptoAPI interface {
-	GetRate(ctx context.Context, opts *GetRateOptions) (float64, error)
+	GetRate(ctx context.Context, fromCurrency, toCurrency string) (float64, error)
+}
+
+// CryptoAPIProviderType represents type of 3rd party provider of crypto API.
+type CryptoAPIProviderType string
+
+const (
+	CryptoAPIProviderCoinbase  CryptoAPIProviderType = "coinbase"
+	CryptoAPIProviderCoinAPI   CryptoAPIProviderType = "coinapi"
+	CryptoAPIProviderCoinGecko CryptoAPIProviderType = "coingecko"
+)
+
+// CryptoAPIProviders is used to store all available crypto APIs.
+type CryptoAPIProviders map[CryptoAPIProviderType]CryptoAPI
+
+// CryptoAPIChain is used to create chain of responsibility for crypto APIs.
+type CryptoAPIChain struct {
+	api  CryptoAPI
+	next *CryptoAPIChain
+}
+
+// NewCryptoAPIChain creates chain of responsibility for crypto APIs.
+// It uses order to determine which API to use first. If no order is provided, default order is used.
+func NewCryptoAPIChain(providers CryptoAPIProviders, order ...CryptoAPIProviderType) (*CryptoAPIChain, error) {
+	var lastProvider *CryptoAPIChain
+
+	// default order
```

Comment:
```suggestion
	if len(order) == 0 {
		order = defaultChainOrder()
	}
```

В ідеалі defaultChainOrder зробити методом CryptoAPIChain, щоб не світити цю функцію на весь package, але lastProvider.defaultChainOrder() більше вводить в оману, ніж допомагає

Reply:
Я трохи переписав, переглянь чи так буде ок будь ласка